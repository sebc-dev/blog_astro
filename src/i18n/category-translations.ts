import type { CategoryTranslations } from "../scripts/category-utils";

/**
 * Centralized category translations for all supported languages
 * This ensures consistency and avoids duplication across the application
 */
export const categoryTranslations = {
  en: {
    framework: "Framework",
    language: "Language", 
    performance: "Performance",
    styling: "Styling",
    backend: "Backend",
    article: "Article"
  } as CategoryTranslations,
  
  fr: {
    framework: "Framework",
    language: "Langage",
    performance: "Performance", 
    styling: "Style",
    backend: "Backend",
    article: "Article"
  } as CategoryTranslations
} as const;

/**
 * Get category translations for a specific language
 * @param lang - The language code (en | fr)
 * @returns The category translations for the specified language
 */
export function getCategoryTranslations(lang: "en" | "fr"): CategoryTranslations {
  return categoryTranslations[lang];
} 