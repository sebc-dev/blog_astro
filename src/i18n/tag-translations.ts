import type { TagTranslations } from "@/lib/article/tag-utils";

/**
 * Centralized tag translations for all supported languages
 * This ensures consistency and avoids duplication across the application
 */
export const tagTranslations = {
  en: {
    guide: "Guide",
    optimization: "Optimization",
    bestPractices: "Best Practices",
    comparison: "Comparison",
  } as TagTranslations,

  fr: {
    guide: "Guide",
    optimization: "Optimisation",
    bestPractices: "Bonnes Pratiques",
    comparison: "Comparaison",
  }
} satisfies Record<"en" | "fr", TagTranslations>;

/**
 * Get tag translations for a specific language
 * @param lang - The language code (en | fr)
 * @returns The tag translations for the specified language
 */
export function getTagTranslations(
  lang: "en" | "fr",
): TagTranslations {
  return tagTranslations[lang];
}