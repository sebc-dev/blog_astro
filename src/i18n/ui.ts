/**
 * Dictionnaires de traduction centralisés pour l'interface utilisateur
 * Langue par défaut : Anglais (en)
 * Langues supportées : Anglais (en), Français (fr)
 */

export const ui = {
  en: {  // LANGUE PAR DÉFAUT
    // Navigation
    'nav.home': 'Home',
    'nav.main': 'Main navigation',
    'nav.articles': 'Articles',
    'nav.blog': 'Blog',
    'nav.about': 'About',
    'nav.contact': 'Contact',
    'nav.services': 'Services',
    'nav.resources': 'Resources',
    // Blog
    'blog.read': 'Read',
    'blog.readMore': 'Read more',
    'blog.publishedOn': 'Published on',
    'blog.author': 'Author',
    'blog.backToBlog': 'Back to Blog',
    'blog.relatedPosts': 'Related Posts',
    'blog.noPostsFound': 'No posts found',
    'blog.searchPlaceholder': 'Search articles...',
    
    // Latest Articles Section
    'latestArticles.title': 'Latest Articles',
    'latestArticles.subtitle': 'Discover our latest insights on web development, modern frameworks and best practices.',
    'latestArticles.noArticles': 'No articles available',
    'latestArticles.noArticlesDesc': 'Articles will be available soon in this language.',
    'latestArticles.viewAll': 'View all articles',
    
    // Categories
    'category.framework': 'Framework',
    'category.language': 'Language',
    'category.performance': 'Performance',
    'category.styling': 'Styling',
    'category.backend': 'Backend',
    'category.article': 'Article',
    
    // Tags
    'tag.guide': 'Guide',
    'tag.optimization': 'Optimization',
    'tag.bestPractices': 'Best Practices',
    'tag.comparison': 'Comparison',
    
    // Interface générale
    'languageSwitcher.label': 'Switch language',
    'languageSwitcher.current': 'Current language',
    'theme.toggle': 'Toggle theme',
    'theme.light': 'Light mode',
    'theme.dark': 'Dark mode',
    
    // Dates et formatage
    'date.format': 'MMMM d, yyyy',
    'date.publishedOn': 'Published on {date}',
    'date.updatedOn': 'Updated on {date}',
    
    // Messages et états
    'loading': 'Loading...',
    'error.pageNotFound': 'Page not found',
    'error.goHome': 'Go to homepage',
    'error.tryAgain': 'Try again',
    
    // Footer
    'footer.allRightsReserved': 'All rights reserved',
    'footer.builtWith': 'Built with',
    
    // SEO et métadonnées
    'meta.description.home': 'Welcome to our blog - Discover articles about web development, technology and more',
    'meta.description.blog': 'Latest articles and tutorials about web development, programming and technology',
    'meta.description.about': 'Learn more about our team and our mission',
    'meta.description.contact': 'Get in touch with us',
  },
  fr: {
    // Navigation
    'nav.home': 'Accueil',
    'nav.main': 'Navigation principale',
    'nav.articles': 'Articles',
    'nav.blog': 'Blog',
    'nav.about': 'À propos',
    'nav.contact': 'Contact',
    'nav.services': 'Services',
    'nav.resources': 'Ressources',
    
    // Blog
    'blog.read': 'Lire',
    'blog.readMore': 'Lire la suite',
    'blog.publishedOn': 'Publié le',
    'blog.author': 'Auteur',
    'blog.backToBlog': 'Retour au Blog',
    'blog.relatedPosts': 'Articles similaires',
    'blog.noPostsFound': 'Aucun article trouvé',
    'blog.searchPlaceholder': 'Rechercher des articles...',
    
    // Latest Articles Section
    'latestArticles.title': 'Derniers Articles',
    'latestArticles.subtitle': 'Découvrez nos dernières réflexions sur le développement web, les frameworks modernes et les meilleures pratiques.',
    'latestArticles.noArticles': 'Aucun article disponible',
    'latestArticles.noArticlesDesc': 'Les articles seront bientôt disponibles dans cette langue.',
    'latestArticles.viewAll': 'Voir tous les articles',
    
    // Categories
    'category.framework': 'Framework',
    'category.language': 'Langage',
    'category.performance': 'Performance',
    'category.styling': 'Style',
    'category.backend': 'Backend',
    'category.article': 'Article',
    
    // Tags
    'tag.guide': 'Guide',
    'tag.optimization': 'Optimisation',
    'tag.bestPractices': 'Bonnes Pratiques',
    'tag.comparison': 'Comparaison',
    
    // Interface générale
    'languageSwitcher.label': 'Changer de langue',
    'languageSwitcher.current': 'Langue actuelle',
    'theme.toggle': 'Basculer le thème',
    'theme.light': 'Mode clair',
    'theme.dark': 'Mode sombre',
    
    // Dates et formatage
    'date.format': 'd MMMM yyyy',
    'date.publishedOn': 'Publié le {date}',
    'date.updatedOn': 'Mis à jour le {date}',
    
    // Messages et états
    'loading': 'Chargement...',
    'error.pageNotFound': 'Page non trouvée',
    'error.goHome': 'Aller à l\'accueil',
    'error.tryAgain': 'Réessayer',
    
    // Footer
    'footer.allRightsReserved': 'Tous droits réservés',
    'footer.builtWith': 'Créé avec',
    
    // SEO et métadonnées
    'meta.description.home': 'Bienvenue sur notre blog - Découvrez des articles sur le développement web, la technologie et plus',
    'meta.description.blog': 'Derniers articles et tutoriels sur le développement web, la programmation et la technologie',
    'meta.description.about': 'En savoir plus sur notre équipe et notre mission',
    'meta.description.contact': 'Contactez-nous',
  },
} as const;

export const defaultLang = 'en'; // ANGLAIS par défaut

/**
 * Type helper pour les clés de traduction
 */
export type UIKeys = keyof typeof ui['en'];

/**
 * Type helper pour les langues supportées
 */
export type Languages = keyof typeof ui; 