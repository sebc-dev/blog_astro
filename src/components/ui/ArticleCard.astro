---
import { getLangFromUrl, useTranslations } from "@/i18n/utils";
import { normalizeCategoryForUrl } from "@/lib/article/category-utils";
import { normalizeTagForUrl } from "@/lib/article/tag-utils";

export interface Props {
  title: string;
  description: string;
  heroImage?: string;
  pubDate: Date;
  category?: string;
  tag?: string;
  slug: string;
  readingTime?: number;
}

const {
  title,
  description,
  heroImage,
  pubDate,
  category = "Article",
  tag,
  slug,
  readingTime = 3,
} = Astro.props;

// Système d'internationalisation
const lang = getLangFromUrl(Astro.url);
const t = useTranslations(lang);

// Formatage de la date
const locale = lang === "fr" ? "fr-FR" : "en-US";
const formattedDate = pubDate.toLocaleDateString(locale, {
  day: "numeric",
  month: "short",
  year: "numeric",
});

// Limite l'extrait pour les cartes
const shortDescription =
  description.length > 120
    ? description.substring(0, 120) + "..."
    : description;

// URLs pour les catégories et tags
const categoryUrl = category ? 
  (lang === "fr" ? `/fr/categorie/${normalizeCategoryForUrl(category)}` : `/category/${normalizeCategoryForUrl(category)}`) : 
  "#";

const tagUrl = tag ? 
  (lang === "fr" ? `/fr/tag/${normalizeTagForUrl(tag)}` : `/tag/${normalizeTagForUrl(tag)}`) : 
  "#";
---

<article
  class="card bg-base-300 border-base-content/10 group h-full overflow-hidden border shadow-lg transition-all duration-300 hover:shadow-xl"
  data-cy="article-card"
  data-article-slug={slug}
>
  <!-- Image de carte -->
  {
    heroImage && (
      <figure class="aspect-video overflow-hidden" data-cy="article-image">
        <img
          src={heroImage}
          alt={title}
          class="h-full w-full object-cover transition-transform duration-300 group-hover:scale-105"
          loading="lazy"
        />
      </figure>
    )
  }

  <div
    class="card-body flex flex-col justify-between p-4"
    data-cy="article-content"
  >
    <!-- Métadonnées compactes -->
    <div class="mb-3 flex flex-wrap gap-1" data-cy="article-metadata">
      {
        category && (
          <a 
            href={categoryUrl} 
            class="badge badge-primary badge-sm hover:badge-primary-focus transition-colors" 
            data-cy="article-category"
          >
            {category}
          </a>
        )
      }
      {
        tag && (
          <a 
            href={tagUrl} 
            class="badge badge-outline badge-sm hover:badge-secondary-focus transition-colors" 
            data-cy="article-tag"
          >
            {tag}
          </a>
        )
      }
      <span
        class="text-secondary-accessible ml-auto text-xs"
        data-cy="article-date"
      >
        {formattedDate}
      </span>
    </div>

    <!-- Titre compact -->
    <h3
      class="card-title group-hover:text-primary mb-3 line-clamp-2 text-lg leading-tight font-bold transition-colors"
      data-cy="article-title"
    >
      <a href={`/blog/${slug}`} class="link-hover" data-cy="article-title-link">
        {title}
      </a>
    </h3>

    <!-- Extrait court avec ellipsis -->
    <p
      class="text-muted-accessible mb-4 line-clamp-2 flex-grow text-sm leading-relaxed"
      data-cy="article-description"
    >
      {shortDescription}
    </p>

    <!-- Footer compact -->
    <div
      class="card-actions mt-auto items-center justify-between"
      data-cy="article-footer"
    >
      <span
        class="text-secondary-accessible text-xs"
        data-cy="article-reading-time"
      >
        {readingTime} min
      </span>

      <a
        href={`/blog/${slug}`}
        class="btn btn-primary btn-xs"
        aria-label={`Lire: ${title}`}
        data-cy="article-read-button"
      >
        {t("blog.read")}
        <svg
          width="12"
          height="12"
          viewBox="0 0 24 24"
          fill="none"
          stroke="currentColor"
          stroke-width="2"
          stroke-linecap="round"
          stroke-linejoin="round"
        >
          <path d="M7 17L17 7"></path>
          <path d="M7 7h10v10"></path>
        </svg>
      </a>
    </div>
  </div>
</article>
