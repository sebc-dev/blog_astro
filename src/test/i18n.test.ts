/**
 * Tests de validation pour le système d'internationalisation (i18n)
 * Vérifie la cohérence des dictionnaires et le bon fonctionnement des utilitaires
 */

import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { ui, defaultLang, type Languages, type UIKeys } from "../i18n/ui";
import {
  getLangFromUrl,
  useTranslations,
  useTranslatedPath,
  getHreflangLinks,
  getPathWithoutLang,
  isValidLang,
  getSupportedLanguages,
  formatDate,
  getLanguageName,
  getLanguageFlag,
} from "../i18n/utils";

// Mock console pour tester les warnings et erreurs
beforeEach(() => {
  vi.spyOn(console, "warn").mockImplementation(() => {});
  vi.spyOn(console, "error").mockImplementation(() => {});
});

afterEach(() => {
  vi.restoreAllMocks();
});

describe("Dictionnaires de traduction (ui.ts)", () => {
  it("devrait avoir l'anglais comme langue par défaut", () => {
    expect(defaultLang).toBe("en");
  });

  it("devrait contenir exactement 2 langues", () => {
    expect(Object.keys(ui)).toHaveLength(2);
    expect(Object.keys(ui)).toContain("en");
    expect(Object.keys(ui)).toContain("fr");
  });

  it("devrait avoir les mêmes clés dans toutes les langues", () => {
    const englishKeys = Object.keys(ui.en).sort();
    const frenchKeys = Object.keys(ui.fr).sort();

    expect(frenchKeys).toEqual(englishKeys);
  });

  it("devrait avoir toutes les clés avec des valeurs non vides", () => {
    Object.entries(ui).forEach(([, translations]) => {
      Object.entries(translations).forEach(([, value]) => {
        expect(value).toBeTruthy();
        expect(typeof value).toBe("string");
        expect(value.trim()).not.toBe("");
      });
    });
  });

  it("devrait contenir les clés essentielles de navigation", () => {
    const essentialNavKeys = [
      "nav.home",
      "nav.blog",
      "nav.about",
      "nav.contact",
      "nav.services",
    ];

    essentialNavKeys.forEach((key) => {
      expect(ui.en).toHaveProperty(key);
      expect(ui.fr).toHaveProperty(key);
    });
  });

  it("devrait contenir les clés essentielles du blog", () => {
    const essentialBlogKeys = [
      "blog.readMore",
      "blog.publishedOn",
      "blog.author",
      "blog.backToBlog",
    ];

    essentialBlogKeys.forEach((key) => {
      expect(ui.en).toHaveProperty(key);
      expect(ui.fr).toHaveProperty(key);
    });
  });

  it("devrait contenir les clés essentielles d'interface", () => {
    const essentialUIKeys = [
      "languageSwitcher.label",
      "theme.toggle",
      "loading",
      "error.pageNotFound",
    ];

    essentialUIKeys.forEach((key) => {
      expect(ui.en).toHaveProperty(key);
      expect(ui.fr).toHaveProperty(key);
    });
  });

  it("devrait contenir les clés de métadonnées SEO", () => {
    const metaKeys = [
      "meta.description.home",
      "meta.description.blog",
      "meta.description.about",
      "meta.description.contact",
    ];

    metaKeys.forEach((key) => {
      expect(ui.en).toHaveProperty(key);
      expect(ui.fr).toHaveProperty(key);
    });
  });
});

describe("Détection de langue (getLangFromUrl)", () => {
  it('devrait retourner "en" pour les URLs sans préfixe de langue', () => {
    const urls = [
      new URL("https://example.com/"),
      new URL("https://example.com/blog"),
      new URL("https://example.com/about"),
      new URL("https://example.com/blog/article-title"),
    ];

    urls.forEach((url) => {
      expect(getLangFromUrl(url)).toBe("en");
    });
  });

  it('devrait retourner "fr" pour les URLs avec préfixe /fr/', () => {
    const urls = [
      new URL("https://example.com/fr/"),
      new URL("https://example.com/fr/blog"),
      new URL("https://example.com/fr/about"),
      new URL("https://example.com/fr/blog/article-titre"),
    ];

    urls.forEach((url) => {
      expect(getLangFromUrl(url)).toBe("fr");
    });
  });

  it("devrait retourner la langue par défaut pour les langues non supportées", () => {
    const urls = [
      new URL("https://example.com/es/blog"),
      new URL("https://example.com/de/about"),
      new URL("https://example.com/invalid/page"),
      new URL("https://example.com/zh/content"),
    ];

    urls.forEach((url) => {
      expect(getLangFromUrl(url)).toBe("en");
    });
  });

  it("devrait gérer les URLs avec des paramètres et fragments", () => {
    const urls = [
      new URL("https://example.com/fr/blog?page=2"),
      new URL("https://example.com/blog#section"),
      new URL("https://example.com/fr/about#contact"),
    ];

    expect(getLangFromUrl(urls[0])).toBe("fr");
    expect(getLangFromUrl(urls[1])).toBe("en");
    expect(getLangFromUrl(urls[2])).toBe("fr");
  });
});

describe("Système de traduction (useTranslations)", () => {
  it("devrait retourner les traductions correctes pour l'anglais", () => {
    const t = useTranslations("en");

    expect(t("nav.home")).toBe("Home");
    expect(t("nav.blog")).toBe("Blog");
    expect(t("blog.readMore")).toBe("Read more");
    expect(t("theme.toggle")).toBe("Toggle theme");
  });

  it("devrait retourner les traductions correctes pour le français", () => {
    const t = useTranslations("fr");

    expect(t("nav.home")).toBe("Accueil");
    expect(t("nav.blog")).toBe("Blog");
    expect(t("blog.readMore")).toBe("Lire la suite");
    expect(t("theme.toggle")).toBe("Basculer le thème");
  });

  it("devrait utiliser le fallback anglais et logger un warning pour les clés manquantes", () => {
    // Simuler une langue avec une clé manquante en utilisant une clé inexistante
    const t = useTranslations("fr");

    // Note: On manipule temporairement l'objet ui pour simuler une traduction manquante
    // Cette approche évite de modifier le fichier source pendant les tests
    // Utiliser vi.mock pour simuler une clé manquante temporairement
    const originalLookup = Object.getOwnPropertyDescriptor(ui.fr, "nav.home");
    Object.defineProperty(ui.fr, "nav.home", {
      value: undefined,
      configurable: true,
    });

    const result = t("nav.home");

    expect(result).toBe("Home"); // Fallback anglais
    expect(console.warn).toHaveBeenCalledWith(
      'Translation missing for key "nav.home" in language "fr". Using fallback from "en".',
    );

    // Restaurer la propriété originale
    if (originalLookup) {
      Object.defineProperty(ui.fr, "nav.home", originalLookup);
    }
  });
  it("devrait retourner la clé elle-même et logger une erreur si aucune traduction n'existe", () => {
    // Test avec une clé qui n'existe nulle part
    const t = useTranslations("en");
    const invalidKey = "nonexistent.key" as UIKeys;
    const result = t(invalidKey);

    expect(result).toBe("nonexistent.key");
    expect(console.error).toHaveBeenCalledWith(
      'Translation missing for key "nonexistent.key" in both "en" and fallback "en".',
    );
  });

  it("devrait gérer le cas où même le fallback anglais est manquant", () => {
    // Simuler l'absence de clé dans les deux langues
    const t = useTranslations("fr");

    // Sauvegarder les descripteurs originaux
    const originalEnDescriptor = Object.getOwnPropertyDescriptor(
      ui.en,
      "nav.home",
    );
    const originalFrDescriptor = Object.getOwnPropertyDescriptor(
      ui.fr,
      "nav.home",
    );

    // Supprimer temporairement les propriétés
    Object.defineProperty(ui.en, "nav.home", {
      value: undefined,
      configurable: true,
    });
    Object.defineProperty(ui.fr, "nav.home", {
      value: undefined,
      configurable: true,
    });

    const result = t("nav.home");

    expect(result).toBe("nav.home"); // Retourne la clé elle-même
    expect(console.error).toHaveBeenCalledWith(
      'Translation missing for key "nav.home" in both "fr" and fallback "en".',
    );

    // Restaurer les propriétés originales
    if (originalEnDescriptor) {
      Object.defineProperty(ui.en, "nav.home", originalEnDescriptor);
    }
    if (originalFrDescriptor) {
      Object.defineProperty(ui.fr, "nav.home", originalFrDescriptor);
    }
  });
});

describe("Gestion des chemins traduits (useTranslatedPath)", () => {
  it("devrait retourner le chemin sans préfixe pour l'anglais", () => {
    const translatePath = useTranslatedPath("en");

    expect(translatePath("/")).toBe("/");
    expect(translatePath("/blog")).toBe("/blog");
    expect(translatePath("/about")).toBe("/about");
    expect(translatePath("/blog/article-title")).toBe("/blog/article-title");
  });

  it("devrait ajouter le préfixe /fr/ pour le français", () => {
    const translatePath = useTranslatedPath("fr");

    expect(translatePath("/", "fr")).toBe("/fr/");
    expect(translatePath("/blog", "fr")).toBe("/fr/blog");
    expect(translatePath("/about", "fr")).toBe("/fr/about");
    expect(translatePath("/blog/article-titre", "fr")).toBe(
      "/fr/blog/article-titre",
    );
  });

  it("devrait gérer les chemins avec ou sans slash initial", () => {
    const translatePath = useTranslatedPath("fr");

    expect(translatePath("blog", "fr")).toBe("/fr/blog");
    expect(translatePath("/blog", "fr")).toBe("/fr/blog");
    expect(translatePath("about/team", "fr")).toBe("/fr/about/team");
    expect(translatePath("/about/team", "fr")).toBe("/fr/about/team");
  });

  it("devrait permettre de spécifier une langue cible différente", () => {
    const translatePath = useTranslatedPath("en");

    expect(translatePath("/blog", "en")).toBe("/blog");
    expect(translatePath("/blog", "fr")).toBe("/fr/blog");

    const translatePathFr = useTranslatedPath("fr");
    expect(translatePathFr("/about", "en")).toBe("/about");
    expect(translatePathFr("/about", "fr")).toBe("/fr/about");
  });

  it("devrait utiliser la langue courante par défaut si aucune langue cible n'est spécifiée", () => {
    const translatePathEn = useTranslatedPath("en");
    const translatePathFr = useTranslatedPath("fr");

    expect(translatePathEn("/blog")).toBe("/blog");
    expect(translatePathFr("/blog")).toBe("/fr/blog");
  });
});

describe("Liens hreflang pour SEO (getHreflangLinks)", () => {
  it("devrait générer les liens hreflang corrects", () => {
    const links = getHreflangLinks("/blog", "https://example.com");

    expect(links).toHaveLength(2);

    const enLink = links.find((link) => link.hreflang === "en");
    const frLink = links.find((link) => link.hreflang === "fr");

    expect(enLink?.href).toBe("https://example.com/blog");
    expect(frLink?.href).toBe("https://example.com/fr/blog");
  });

  it("devrait fonctionner sans URL de base", () => {
    const links = getHreflangLinks("/about");

    expect(links).toHaveLength(2);

    const enLink = links.find((link) => link.hreflang === "en");
    const frLink = links.find((link) => link.hreflang === "fr");

    expect(enLink?.href).toBe("/about");
    expect(frLink?.href).toBe("/fr/about");
  });

  it("devrait gérer les chemins sans slash initial", () => {
    const links = getHreflangLinks("contact", "https://example.com");

    expect(links).toHaveLength(2);

    const enLink = links.find((link) => link.hreflang === "en");
    const frLink = links.find((link) => link.hreflang === "fr");

    expect(enLink?.href).toBe("https://example.com/contact");
    expect(frLink?.href).toBe("https://example.com/fr/contact");
  });

  it("devrait gérer la page d'accueil", () => {
    const links = getHreflangLinks("/", "https://example.com");

    expect(links).toHaveLength(2);

    const enLink = links.find((link) => link.hreflang === "en");
    const frLink = links.find((link) => link.hreflang === "fr");

    expect(enLink?.href).toBe("https://example.com/");
    expect(frLink?.href).toBe("https://example.com/fr/");
  });
});

describe("Extraction du chemin sans langue (getPathWithoutLang)", () => {
  it("devrait extraire le chemin sans préfixe de langue française", () => {
    const urls = [
      { url: new URL("https://example.com/fr/blog"), expected: "/blog" },
      { url: new URL("https://example.com/fr/about"), expected: "/about" },
      { url: new URL("https://example.com/fr/"), expected: "/" },
      {
        url: new URL("https://example.com/fr/blog/article-titre"),
        expected: "/blog/article-titre",
      },
    ];

    urls.forEach(({ url, expected }) => {
      expect(getPathWithoutLang(url)).toBe(expected);
    });
  });

  it("devrait retourner le chemin complet pour les URLs anglaises", () => {
    const urls = [
      { url: new URL("https://example.com/blog"), expected: "/blog" },
      { url: new URL("https://example.com/about"), expected: "/about" },
      { url: new URL("https://example.com/"), expected: "/" },
      {
        url: new URL("https://example.com/blog/article-title"),
        expected: "/blog/article-title",
      },
    ];

    urls.forEach(({ url, expected }) => {
      expect(getPathWithoutLang(url)).toBe(expected);
    });
  });

  it("devrait gérer les URLs avec des langues non supportées", () => {
    const urls = [
      { url: new URL("https://example.com/es/blog"), expected: "/es/blog" },
      { url: new URL("https://example.com/de/about"), expected: "/de/about" },
      {
        url: new URL("https://example.com/invalid/page"),
        expected: "/invalid/page",
      },
    ];

    urls.forEach(({ url, expected }) => {
      expect(getPathWithoutLang(url)).toBe(expected);
    });
  });

  it("devrait gérer les chemins vides correctement", () => {
    const urls = [
      { url: new URL("https://example.com/fr"), expected: "/" },
      { url: new URL("https://example.com/fr/"), expected: "/" },
    ];

    urls.forEach(({ url, expected }) => {
      expect(getPathWithoutLang(url)).toBe(expected);
    });
  });
});

describe("Validation des langues (isValidLang)", () => {
  it("devrait valider les langues supportées", () => {
    expect(isValidLang("en")).toBe(true);
    expect(isValidLang("fr")).toBe(true);
  });

  it("devrait rejeter les langues non supportées", () => {
    expect(isValidLang("es")).toBe(false);
    expect(isValidLang("de")).toBe(false);
    expect(isValidLang("invalid")).toBe(false);
    expect(isValidLang("")).toBe(false);
    expect(isValidLang("EN")).toBe(false); // Sensible à la casse
    expect(isValidLang("FR")).toBe(false); // Sensible à la casse
  });

  it("devrait retourner false pour des valeurs nulles ou undefined", () => {
    // @ts-expect-error - Test intentionnel avec des valeurs incorrectes
    expect(isValidLang(null)).toBe(false);
    // @ts-expect-error - Test intentionnel avec des valeurs incorrectes
    expect(isValidLang(undefined)).toBe(false);
  });
});

describe("Langues supportées (getSupportedLanguages)", () => {
  it("devrait retourner toutes les langues supportées", () => {
    const languages = getSupportedLanguages();

    expect(languages).toHaveLength(2);
    expect(languages).toContain("en");
    expect(languages).toContain("fr");
  });

  it("devrait retourner un tableau ordonné", () => {
    const languages = getSupportedLanguages();
    const sortedLanguages = [...languages].sort();

    expect(languages).toEqual(expect.arrayContaining(sortedLanguages));
  });
});

describe("Formatage des dates (formatDate)", () => {
  const testDate = new Date("2024-01-15");

  it("devrait formater les dates en anglais", () => {
    const formatted = formatDate(testDate, "en");
    // Vérifier la structure plutôt que le contenu exact
    expect(formatted).toContain("2024");
    expect(formatted).toContain("15");
    expect(formatted).toContain("January");
    // Test de la structure complète
    expect(formatted).toMatch(/^[A-Za-z]+ \d{1,2}, \d{4}$/);
  });

  it("devrait formater les dates en français", () => {
    const formatted = formatDate(testDate, "fr");
    // Vérifier la structure plutôt que le contenu exact
    expect(formatted).toContain("2024");
    expect(formatted).toContain("15");
    expect(formatted).toContain("janvier");
    // Test de la structure complète pour le français
    expect(formatted).toMatch(/^\d{1,2} [a-zA-Zâêîôûàèùç]+ \d{4}$/);
  });

  it("devrait gérer différentes dates", () => {
    const dates = [
      new Date("2023-12-25"), // Noël
      new Date("2024-07-14"), // Fête nationale française
      new Date("2024-02-29"), // Année bissextile
    ];

    dates.forEach((date) => {
      const formattedEn = formatDate(date, "en");
      const formattedFr = formatDate(date, "fr");

      expect(formattedEn).toBeTruthy();
      expect(formattedFr).toBeTruthy();
      expect(formattedEn).not.toBe(formattedFr);
    });
  });

  it("devrait utiliser le bon format de locale", () => {
    const formatted = formatDate(testDate, "fr");
    // Le format français utilise "janvier" et non "January"
    expect(formatted).toContain("janvier");
    expect(formatted).not.toContain("January");
  });

  it("devrait maintenir un formatage consistant (test de snapshot)", () => {
    // Test de snapshot plus robuste avec date fixe et timezone normalisée
    const fixedDate = new Date("2024-01-15T12:00:00.000Z");

    // Test pour l'anglais
    const formattedEn = formatDate(fixedDate, "en");
    expect(formattedEn).toMatchSnapshot("formatDate-en-2024-01-15");

    // Test pour le français
    const formattedFr = formatDate(fixedDate, "fr");
    expect(formattedFr).toMatchSnapshot("formatDate-fr-2024-01-15");

    // Vérification que les formats sont différents
    expect(formattedEn).not.toBe(formattedFr);
  });

  it("devrait gérer différentes dates de manière cohérente", () => {
    const testDates = [
      { date: new Date("2024-01-01T12:00:00.000Z"), label: "Nouvel An" },
      {
        date: new Date("2024-07-14T12:00:00.000Z"),
        label: "Fête nationale française",
      },
      { date: new Date("2024-12-25T12:00:00.000Z"), label: "Noël" },
    ];

    testDates.forEach(({ date, label }) => {
      const formattedEn = formatDate(date, "en");
      const formattedFr = formatDate(date, "fr");

      // Vérifier que le formatage produit quelque chose
      expect(formattedEn, `Date en anglais pour ${label}`).toBeTruthy();
      expect(formattedFr, `Date en français pour ${label}`).toBeTruthy();

      // Vérifier que les formats sont différents
      expect(formattedEn, `Formats différents pour ${label}`).not.toBe(
        formattedFr,
      );

      // Vérifier la présence de l'année
      expect(formattedEn, `Année présente en anglais pour ${label}`).toContain(
        "2024",
      );
      expect(formattedFr, `Année présente en français pour ${label}`).toContain(
        "2024",
      );
    });
  });
});

describe("Noms des langues (getLanguageName)", () => {
  it("devrait retourner les noms corrects des langues", () => {
    expect(getLanguageName("en")).toBe("English");
    expect(getLanguageName("fr")).toBe("Français");
  });

  it("devrait retourner le code de langue comme fallback pour une langue non configurée", () => {
    // @ts-expect-error - Test avec une langue non supportée
    expect(getLanguageName("es")).toBe("es");
    // @ts-expect-error - Test avec une langue non supportée
    expect(getLanguageName("de")).toBe("de");
  });
});

describe("Drapeaux des langues (getLanguageFlag)", () => {
  it("devrait retourner les drapeaux corrects", () => {
    expect(getLanguageFlag("en")).toBe("🇺🇸");
    expect(getLanguageFlag("fr")).toBe("🇫🇷");
  });

  it("devrait retourner un drapeau générique pour une langue non configurée", () => {
    // @ts-expect-error - Test avec une langue non supportée
    expect(getLanguageFlag("es")).toBe("🌐");
    // @ts-expect-error - Test avec une langue non supportée
    expect(getLanguageFlag("de")).toBe("🌐");
  });
});

describe("Cohérence globale du système i18n", () => {
  it("devrait avoir une architecture cohérente", () => {
    // Vérifier que tous les éléments du système sont en place
    expect(ui).toBeDefined();
    expect(defaultLang).toBeDefined();
    expect(typeof getLangFromUrl).toBe("function");
    expect(typeof useTranslations).toBe("function");
    expect(typeof useTranslatedPath).toBe("function");
  });

  it("devrait supporter l'ajout facile de nouvelles langues", () => {
    // Vérifier que la structure permet l'extension
    const languages = Object.keys(ui);
    expect(languages.length).toBeGreaterThan(0);

    // Toutes les langues devraient avoir le même ensemble de clés
    const referenceKeys = Object.keys(ui[languages[0] as Languages]);
    languages.forEach((lang) => {
      const langKeys = Object.keys(ui[lang as Languages]);
      expect(langKeys.sort()).toEqual(referenceKeys.sort());
    });
  });

  it("devrait maintenir la cohérence des types", () => {
    // Vérifier que les types sont cohérents
    const languages = getSupportedLanguages();
    languages.forEach((lang) => {
      expect(isValidLang(lang)).toBe(true);
      expect(getLanguageName(lang)).toBeTruthy();
      expect(getLanguageFlag(lang)).toBeTruthy();
    });
  });

  it("devrait avoir des traductions complètes pour toutes les fonctionnalités", () => {
    const essentialKeys = [
      "nav.home",
      "nav.blog",
      "nav.about",
      "blog.readMore",
      "blog.publishedOn",
      "theme.toggle",
      "loading",
      "error.pageNotFound",
      "meta.description.home",
    ];

    getSupportedLanguages().forEach((lang) => {
      const t = useTranslations(lang);
      essentialKeys.forEach((key) => {
        const translation = t(key as UIKeys);
        expect(translation).toBeTruthy();
        expect(translation).not.toBe(key); // Pas de fallback vers la clé
      });
    });
  });
});
