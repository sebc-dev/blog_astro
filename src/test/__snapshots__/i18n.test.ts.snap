// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Formatage des dates (formatDate) > devrait maintenir un formatage consistant (test de snapshot) > formatDate-en-2024-01-15 1`] = `"January 15, 2024"`;

exports[`Formatage des dates (formatDate) > devrait maintenir un formatage consistant (test de snapshot) > formatDate-fr-2024-01-15 1`] = `"15 janvier 2024"`;
